src/Data/String/CodePoints.purs

```diff
@@ -0,0 +1,216 @@
+  , length
+  , singleton
+  , splitAt
+  , take
+  , takeWhile
+  , toCodePointArray
+  , uncons
+  ) where
+
+import Data.Array as Array
+import Data.Char as Char
+import Data.List (List(Cons, Nil), fromFoldable)
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.Newtype (class Newtype)
+import Data.String as String
+-- WARN: This list must be updated to re-export any exports added to Data.String. That makes me sad.
+import Data.String (Pattern(..), Replacement(..), charAt, charCodeAt, contains, fromCharArray, joinWith, localeCompare, null, replace, replaceAll, split, stripPrefix, stripSuffix, toChar, toCharArray, toLower, toUpper, trim) as StringReExports
+import Data.Tuple (Tuple(Tuple))
+import Data.Unfoldable (unfoldr)
+import Prelude (class Eq, class Ord, (&&), (||), (*), (+), (-), (<$>), (<), (<=), (<<<), (/), (<>), (==), mod)
+
+
+newtype CodePoint = CodePoint Int
+
+derive instance eqCodePoint :: Eq CodePoint
+derive instance ordCodePoint :: Ord CodePoint
+derive instance newtypeCodePoint :: Newtype CodePoint _
+
+codePointFromInt :: Int -> Maybe CodePoint
+codePointFromInt n | 0 <= n && n <= 0x10FFFF = Just (CodePoint n)
+codePointFromInt n = Nothing
+
+codePointToInt :: CodePoint -> Int
+codePointToInt (CodePoint n) = n
+
+codePointFromSurrogatePair :: Int -> Int -> Maybe CodePoint
+codePointFromSurrogatePair lead trail | isLead lead && isTrail trail =
+  Just (unsurrogate lead trail)
+codePointFromSurrogatePair _ _ = Nothing
+
+unsurrogate :: Int -> Int -> CodePoint
+unsurrogate h l = CodePoint ((h - 0xD800) * 0x400 + (l - 0xDC00) + 0x10000)
+
+isLead :: Int -> Boolean
+isLead cu = 0xD800 <= cu && cu <= 0xDBFF
+
+isTrail :: Int -> Boolean
+isTrail cu = 0xDC00 <= cu && cu <= 0xDFFF
+
+fromCharCode :: Int -> String
+fromCharCode = String.singleton <<< Char.fromCharCode
+
+unsafeCodePointAt0 :: String -> CodePoint
+unsafeCodePointAt0 = _unsafeCodePointAt0 unsafeCodePointAt0Fallback
+
+foreign import _unsafeCodePointAt0
+  :: (String -> CodePoint)
+  -> String
+  -> CodePoint
+
+unsafeCodePointAt0Fallback :: String -> CodePoint
+unsafeCodePointAt0Fallback s | String.length s == 1 = CodePoint (_unsafeCharCodeAt 0 s)
+unsafeCodePointAt0Fallback s = CodePoint (((lead - 0xD800) * 0x400) + (trail - 0xDC00) + 0x10000)
+  where
+    lead = _unsafeCharCodeAt 0 s
+    trail = _unsafeCharCodeAt 1 s
+
+foreign import _unsafeCharCodeAt :: Int -> String -> Int
+
+
+codePointAt :: Int -> String -> Maybe CodePoint
+codePointAt 0 "" = Nothing
+codePointAt 0 s = Just (unsafeCodePointAt0 s)
+codePointAt n s = _codePointAt codePointAtFallback Just Nothing unsafeCodePointAt0 n s
+
+foreign import _codePointAt
+  :: (Int -> String -> Maybe CodePoint)
+  -> (forall a. a -> Maybe a)
+  -> (forall a. Maybe a)
+  -> (String -> CodePoint)
+  -> Int
+  -> String
+  -> Maybe CodePoint
+
+codePointAtFallback :: Int -> String -> Maybe CodePoint
+codePointAtFallback n s = Array.index (toCodePointArray s) n
+
+
+count :: (CodePoint -> Boolean) -> String -> Int
```

Comment:
Why does `count` require a predicate? Is it common to consider only certain characters of a code set when counting? Or is it because code points have complications, like digraphs?